src/core_plugins/kibana/public/home/components/tutorial_directory.js

```diff
@@ -0,0 +1,119 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Synopsis } from './synopsis';
+import {
+  KuiTabs,
+  KuiTab,
+  KuiFlexItem,
+  KuiFlexGrid,
+} from 'ui_framework/components';
+import { getTutorials } from '../load_tutorials';
+
+const ALL = 'all';
+
+export class TutorialDirectory extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.tabs = [{
+      id: ALL,
+      name: 'All',
+    }, {
+      id: 'logging',
+      name: 'Logging',
+    }, {
+      id: 'metrics',
+      name: 'Metrics',
+    }, {
+      id: 'security',
+      name: 'Security Analytics',
+    }];
+
+    let openTab = ALL;
+    if (props.openTab && this.tabs.some(tab => { return tab.id === props.openTab; })) {
+      openTab = props.openTab;
+    }
+    this.state = {
+      selectedTabId: openTab,
+      tutorials: []
+    };
+  }
+
+  async componentWillMount() {
+    const tutorials = await getTutorials();
```

Comment:
Asyncs in componentWillMount can cause issues.  e.g. I can simulate a failure if i throw in some long running loops into the async call, and in the mean time leave the page.

<img width="1258" alt="screen shot 2017-12-18 at 12 13 38 pm" src="https://user-images.githubusercontent.com/16563603/34118529-151a886a-e3ed-11e7-9e86-740006001dd5.png">

It's because you could get back from `getTutorials` after the component has unmounted and then there is no "this".  So you need to manually track whether the component is mounted via some kind of isMounted flag and unset it in componentWillUnmount.

In additional, apparently a better place to set async calls is componentDidMount not componentWillMount, bc componentWillMount could be called multiple times: https://daveceddia.com/where-fetch-data-componentwillmount-vs-componentdidmount/

Reply:
Looks like setting state in componentDidMount is an antipattern too. Lint returned the following error `49:7  error  Do not use setState in componentDidMount  react/no-did-mount-set-state`

Reply:
Yea, ideally we wouldn't have either, but I think the only way to avoid it is to use redux.  I went through this when I introduced react-grid-layout in dashboard, before I added redux: https://github.com/elastic/kibana/pull/13853/files#diff-ca84256adcfe19a824a1a489a6dc5ddeR29

I used an eslint disable line so it could go in componentDidMount, and manually tracked whether the component is mounted.